Events: Fix fixed pcie wake event

lvjianmin-loongson opened the pull request:
Some platforms support the PCIEXP_WAKE_STS bit in the PM1 Status register and the PCIEXP_WAKE_EN bit in the PM1 Enable register. But ACPI_FADT_PCI_EXPRESS_WAKE is not set in fixed feature flag of FADT, skip handling it during processing fixed events if no ACPI_FADT_PCI_EXPRESS_WAKE set.

Second, fixed pcie wakeup event should be enabled before entering sleep state just as GPEs with wakeup ability, or the system will be not waked up by pcie devices due to its disabled state, and it should be cleared on waking up so that the event is disabled at runtime (it is not a runtime event).

Fixes: 32d875705c8e ("Events: Support fixed pcie wake event")

sacdintel commented:
Hi Jianmin,
Has this PR been tested on Unix-like(or Linux) and Windows/MSVC environments? Would it also address the concerns/resolutions found in PR #818 that we were not sure how to proceed with? I believe the PCIe Wake fixed event enable and disable support was added by Loongson before, possible sometime last year (2022) or so. Thanks!
- Saket Dumbre

sacdintel commented:
I believe it was PR #757 and upon discussing it, the Windows implementation of ACPI still permanently disables the PCIe wake (fixed) event enable/disable support, so you might want to add it there as well after approval from MSFT or ASWG/USWG (I am not sure exactly who or where though)

lvjianmin-loongson commented:
> ACPI_EVENT_PCIE_WAKE

lvjianmin-loongson commented:
> Hi Jianmin, Has this PR been tested on Unix-like(or Linux) and Windows/MSVC environments? Would it also address the concerns/resolutions found in PR #818 that we were not sure how to proceed with? I believe the PCIe Wake fixed event enable and disable support was added by Loongson before, possible sometime last year (2022) or so. Thanks!
> 
> * Saket Dumbre

Yes, it addressed issues mentioned in PR#818, and has been tested for many machines, please see
https://bugzilla.kernel.org/show_bug.cgi?id=217069

lvjianmin-loongson commented on source/components/events/evxfevnt.c:
```diff
@@ -320,13 +320,18 @@ AcpiEnableEvent (
 
+    if (Event == ACPI_EVENT_PCIE_WAKE &&
+        !(AcpiGbl_FADT.Flags & ACPI_FADT_PCI_EXPRESS_WAKE)) {
+            return_ACPI_STATUS (AE_BAD_PARAMETER);
```
Good suggestion, it seems that AE_SUPPORT is more better here and others in the patch, I'll change them.
Thanks!

lvjianmin-loongson commented:
> I believe it was PR #757 and upon discussing it, the Windows implementation of ACPI still permanently disables the PCIe wake (fixed) event enable/disable support, so you might want to add it there as well after approval from MSFT or ASWG/USWG (I am not sure exactly who or where though)

Ok, I think I can discuss it with ASWG first.
Thanks!

rafaeljw commented:
> Some platforms support the PCIEXP_WAKE_STS bit in the PM1 Status register and the PCIEXP_WAKE_EN bit in the PM1 Enable register. But ACPI_FADT_PCI_EXPRESS_WAKE is not set in fixed feature flag of FADT, skip handling it during processing fixed events if no ACPI_FADT_PCI_EXPRESS_WAKE set.
> 
> Second, fixed pcie wakeup event should be enabled before entering sleep state just as GPEs with wakeup ability, or the system will be not waked up by pcie devices due to its disabled state, and it should be cleared on waking up so that the event is disabled at runtime (it is not a runtime event).
> 
> Fixes: [32d8757](https://github.com/acpica/acpica/commit/32d875705c8ee8f99fd8b78dbed48633486a7640) ("Events: Support fixed pcie wake event")

Can you please send the corresponding Linux patch to linux-acpi@vger.kernel.org?

lvjianmin-loongson commented:
> > I believe it was PR #757 and upon discussing it, the Windows implementation of ACPI still permanently disables the PCIe wake (fixed) event enable/disable support, so you might want to add it there as well after approval from MSFT or ASWG/USWG (I am not sure exactly who or where though)
> 
> Ok, I think I can discuss it with ASWG first. Thanks!

I have disscussed it with ASWG, as described in ACPI spec 6.5,
4.8.3.1.1 PM1 Status Registers, some fixed hardware features such
as TMR, GBL, PWRBTN, SLPBTN and RTC, will raise an interrupt event
if related xx_EN and xx_STS are set. But PCIEXP_WAKE as one kind
of fixed hardware feature, which does not generate an SCI, and is
used only as a wake event.

So, for PCIEXP_WAKE, we don't have to handle it in SCI interrupt
and disable/enable during S3/S4 period.

lvjianmin-loongson commented:
> > Some platforms support the PCIEXP_WAKE_STS bit in the PM1 Status register and the PCIEXP_WAKE_EN bit in the PM1 Enable register. But ACPI_FADT_PCI_EXPRESS_WAKE is not set in fixed feature flag of FADT, skip handling it during processing fixed events if no ACPI_FADT_PCI_EXPRESS_WAKE set.
> > Second, fixed pcie wakeup event should be enabled before entering sleep state just as GPEs with wakeup ability, or the system will be not waked up by pcie devices due to its disabled state, and it should be cleared on waking up so that the event is disabled at runtime (it is not a runtime event).
> > Fixes: [32d8757](https://github.com/acpica/acpica/commit/32d875705c8ee8f99fd8b78dbed48633486a7640) ("Events: Support fixed pcie wake event")
> 
> Can you please send the corresponding Linux patch to [linux-acpi@vger.kernel.org](mailto:linux-acpi@vger.kernel.org)?

As I replied above, should we provide a fixed patch or revert?

rafaeljw commented:
> > > Some platforms support the PCIEXP_WAKE_STS bit in the PM1 Status register and the PCIEXP_WAKE_EN bit in the PM1 Enable register. But ACPI_FADT_PCI_EXPRESS_WAKE is not set in fixed feature flag of FADT, skip handling it during processing fixed events if no ACPI_FADT_PCI_EXPRESS_WAKE set.
> > > Second, fixed pcie wakeup event should be enabled before entering sleep state just as GPEs with wakeup ability, or the system will be not waked up by pcie devices due to its disabled state, and it should be cleared on waking up so that the event is disabled at runtime (it is not a runtime event).
> > > Fixes: [32d8757](https://github.com/acpica/acpica/commit/32d875705c8ee8f99fd8b78dbed48633486a7640) ("Events: Support fixed pcie wake event")
> > 
> > 
> > Can you please send the corresponding Linux patch to [linux-acpi@vger.kernel.org](mailto:linux-acpi@vger.kernel.org)?
> 
> As I replied above, should we provide a fixed patch or revert?

Yes, please.

There is a functional regression on multiple systems because of this issue.

lvjianmin-loongson commented:
> > > > Some platforms support the PCIEXP_WAKE_STS bit in the PM1 Status register and the PCIEXP_WAKE_EN bit in the PM1 Enable register. But ACPI_FADT_PCI_EXPRESS_WAKE is not set in fixed feature flag of FADT, skip handling it during processing fixed events if no ACPI_FADT_PCI_EXPRESS_WAKE set.
> > > > Second, fixed pcie wakeup event should be enabled before entering sleep state just as GPEs with wakeup ability, or the system will be not waked up by pcie devices due to its disabled state, and it should be cleared on waking up so that the event is disabled at runtime (it is not a runtime event).
> > > > Fixes: [32d8757](https://github.com/acpica/acpica/commit/32d875705c8ee8f99fd8b78dbed48633486a7640) ("Events: Support fixed pcie wake event")
> > > 
> > > 
> > > Can you please send the corresponding Linux patch to [linux-acpi@vger.kernel.org](mailto:linux-acpi@vger.kernel.org)?
> > 
> > 
> > As I replied above, should we provide a fixed patch or revert?
> 
> Yes, please.
> 
> There is a functional regression on multiple systems because of this issue.

Should I(myself) to submit the reverted patch?  to both linux and ACPICA?

lvjianmin-loongson commented:
The patch in this PR has not been needed any more, so close it.
